File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/connection/polling/WaitForConnectionActivity.java
package com.lasthopesoftware.bluewater.client.connection.polling;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import com.lasthopesoftware.bluewater.R;
import com.lasthopesoftware.bluewater.client.connection.IConnectionProvider;
import com.lasthopesoftware.bluewater.settings.ApplicationSettingsActivity;
import com.namehillsoftware.handoff.promises.Promise;
import com.namehillsoftware.handoff.promises.response.VoidResponse;

import java.util.concurrent.CancellationException;

public class WaitForConnectionActivity extends Activity {

	public static void beginWaiting(final Context context, final Runnable onConnectionRegainedListener) {
		context.startActivity(new Intent(context, WaitForConnectionActivity.class));
		PollConnectionService.pollSessionConnection(context)
			.then(c -> {
				onConnectionRegainedListener.run();
				return null;
			});
	}
	
	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_wait_for_connection);
		
		final Intent selectServerIntent = new Intent(this, ApplicationSettingsActivity.class);

		final Promise<IConnectionProvider> pollSessionConnection = PollConnectionService.pollSessionConnection(this);

		findViewById(R.id.btnCancel).setOnClickListener(v -> {
			pollSessionConnection.cancel();
			startActivity(selectServerIntent);
			finish();
		});

		pollSessionConnection
			.then(
				new VoidResponse<>(c -> finish()),
				new VoidResponse<>(e -> {
					if (e instanceof CancellationException) startActivity(selectServerIntent);

					finish();
				}));
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/connection/polling/WaitForConnectionActivity.kt
package com.lasthopesoftware.bluewater.client.connection.polling

import android.content.Context
import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.lasthopesoftware.bluewater.ActivityApplicationNavigation
import com.lasthopesoftware.bluewater.R
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.connection.polling.PollConnectionService.Companion.pollSessionConnection
import com.lasthopesoftware.bluewater.shared.MagicPropertyBuilder
import com.lasthopesoftware.bluewater.shared.android.intents.IntentBuilder
import com.lasthopesoftware.bluewater.shared.android.intents.getIntent
import com.lasthopesoftware.bluewater.shared.android.intents.safelyGetParcelableExtra
import java.util.concurrent.CancellationException

class WaitForConnectionActivity : AppCompatActivity() {
	private val applicationNavigation by lazy {
		ActivityApplicationNavigation(
			this,
			IntentBuilder(this),
		)
	}

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_wait_for_connection)

		val libraryId = intent.safelyGetParcelableExtra<LibraryId>(libraryIdProperty) ?: return

		val pollSessionConnection = pollSessionConnection(this, libraryId)

		findViewById<View>(R.id.btnCancel).setOnClickListener {
			pollSessionConnection.cancel()
			applicationNavigation.viewApplicationSettings()
			finish()
		}

		pollSessionConnection
			.then(
				{ finish() },
				{ e ->
					if (e is CancellationException) applicationNavigation.viewApplicationSettings()
					finish()
				})
	}

	companion object {
		private val libraryIdProperty by lazy { MagicPropertyBuilder.buildMagicPropertyName<WaitForConnectionActivity>("libraryId") }

		fun beginWaiting(context: Context, libraryId: LibraryId, onConnectionRegainedListener: Runnable) {
			val intent = context.getIntent<WaitForConnectionActivity>().apply {
				putExtra(libraryIdProperty, libraryId)
			}

			context.startActivity(intent)
			pollSessionConnection(context, libraryId).then { onConnectionRegainedListener.run() }
		}
	}
}
